File: src/simulation/aircraft.ts
import type { AirOrder as NativeAirOrder } from '../multiplayer/generated/AirOrder';
import { physicalLoss } from '../game/session/battleRules';
import type { ShipDefinition, Vec3 } from '../ships/blueprint';
import type { FleetActor, Team } from './battle';
import type { CombatEvent } from './combat';
import type { Shell } from './damage';
import { add, clamp, length, localToWorld, scale, sub, worldToLocal } from '../game/geometry';
import { equipmentCondition } from '../game/machinery';
import { motionVelocity } from '../game/session/motion';
import { type Torpedo } from '../game/torpedoAim';
import { initialFlightControls, TAKEOFF_ROLL_SECONDS } from './aircraftFlight';
import { initialAirPilot, type AirPilot } from './aircraftTactics';
import { aircraftDeckAttitude, aircraftGroundPose } from '../game/aircraftGroundPose';
import { DEFAULT_AIR_TORPEDO } from '../ships/aircraftWeapons';

/** Declared with the frame, in Rust (`naval_sim::aircraft`); the engine's
 * plane keeps its pilot, which the frame publishes as `behavior`. */
export type { FlightPhase, AirFlight, DeckStatus, AirRelease } from '../game/session/elements';
export type AirOrder = NativeAirOrder;
import type { Aircraft as SessionAircraft, AirWingState as SessionAirWingState, FleetActor as SessionActor, AirFlight, AirRelease } from '../game/session/elements';
import { hasFoldingWings, AIRCRAFT_ENDURANCE_SECONDS, FIGHTER_AMMO_BURSTS, AIRCRAFT_REPAIR_HP, aircraftServiceSeconds, deckClearance, flightSize, deckCapacity, terminalAircraft, activeFlight, airborne, aircraftDeckSpot, onFlightDeck, airServiceAvailable, squadronFlights, recoveryQueue } from '../game/airWing';
export { hasFoldingWings, AIRCRAFT_ENDURANCE_SECONDS, FIGHTER_AMMO_BURSTS, AIRCRAFT_REPAIR_HP, aircraftServiceSeconds, deckClearance, flightSize, deckCapacity, terminalAircraft, activeFlight, airborne, aircraftDeckSpot, onFlightDeck, airServiceAvailable, squadronFlights, recoveryQueue } from '../game/airWing';
export interface Aircraft extends SessionAircraft { kills: number; pilot: AirPilot; }
export interface AirWingState extends SessionAirWingState { planes: Aircraft[]; }
function deckPose(p: Aircraft, actor: FleetActor, local: Vec3) {
  p.deckPosition = [...local]; p.position = localToWorld(local, actor.motion);
  p.deckHeading = 0;
  Object.assign(p, aircraftDeckAttitude(actor.motion, p.modelId));
  p.velocity = motionVelocity(actor.motion);
}
function taxi(p: Aircraft, actor: FleetActor, destination: Vec3, speed: number, dt: number): boolean {
  const current = p.deckPosition ?? worldToLocal(p.position, actor.motion);
  const delta = sub(destination, current), distance = length(delta);
  const local = add(current, scale(delta, Math.min(1, speed * dt / (distance || 1))));
  deckPose(p, actor, local);
  if (distance > .1) {
    p.deckHeading = Math.atan2(delta[0], -delta[2]);
    Object.assign(p, aircraftDeckAttitude(actor.motion, p.modelId, p.deckHeading));
  }
  return distance <= speed * dt;
}
export function createAirWing(def: ShipDefinition, ownerId: string, team: Team): AirWingState | undefined {
  if (!def.airWing) return;
  const state: AirWingState = { launchCooldown: 0, flights: [], flightSequence: 0, transferCooldown: 0, planes: def.airWing.squadrons.flatMap(s => Array.from({ length: s.count }, (_, i) => ({
    id: `${ownerId}/${s.id}/${i + 1}`, ownerId, team, squadronId: s.id, modelId: s.modelId, role: s.role,
    phase: 'ready' as const, wingFold: hasFoldingWings(s.modelId) ? 1 : 0, position: [0, 0, 0] as Vec3, previousPosition: [0, 0, 0] as Vec3, velocity: [0, 0, 0] as Vec3,
    heading: 0, pitch: 0, bank: 0, hp: 100, ammo: s.role === 'fighter' ? FIGHTER_AMMO_BURSTS : 0, payload: s.role !== 'fighter', timer: 0, flightTime: 0, cooldown: 0, kills: 0,
    controls: initialFlightControls(), pilot: initialAirPilot(),
  }))) };
  return state;
}
function validAirOrder(actor: FleetActor, flightId: string, planes: Aircraft[], order: AirOrder, actors: FleetActor[]) {
  if (order.kind === 'return') return true;
  if (order.kind === 'strike' || order.kind === 'intercept-contact' || order.kind === 'search-area') return false; // Rust observation-mode orders.
  if (!planes.length || planes.some(p => p.flightTime > 470 || p.hp < 25)) return false;
  if (order.kind === 'patrol') return order.point.length === 3 && order.point.every(Number.isFinite)
    && Math.hypot(order.point[0] - actor.motion.x, order.point[2] - actor.motion.z) <= 30000;
  if (order.kind === 'attack') {
    const target = actors.find(a => a.motion.id === order.targetId);
    return !!target && target.team !== actor.team && !physicalLoss(target) && target.motion.y >= -8
      && planes.every(p => p.role !== 'fighter' && p.payload);
  }
  if (planes.some(p => p.role !== 'fighter' || p.ammo <= 0)) return false;
  if (order.kind === 'defend') return !order.targetId || actors.some(a => a.motion.id === order.targetId && a.team === actor.team && !physicalLoss(a));
  return order.flightId !== flightId && actors.some(a => (order.kind === 'escort' ? a.team === actor.team : a.team !== actor.team)
    && a.airWing?.flights.some(f => f.id === order.flightId && activeFlight(f, a.airWing!.planes)));
}
export function launchSquadron(actor: FleetActor, squadronId: string, target?: FleetActor, requestedOrder?: AirOrder, actors: FleetActor[] = target ? [actor, target] : [actor], flightId?: string): number {
  if (!airServiceAvailable(actor)) return 0;
  const state = actor.airWing!;
  if (state.flights.filter(f => activeFlight(f, state.planes)).length >= actor.definition.airWing!.maxActiveFlights) return 0;
  const flight = squadronFlights(actor).find(f => f.squadronId === squadronId && (!flightId || f.id === flightId)
    && state.planes.filter(p => f.planeIds.includes(p.id) && p.phase !== 'lost').length > 0
    && state.planes.filter(p => f.planeIds.includes(p.id)).every(p => p.phase === 'ready' || p.phase === 'lost'));
  if (!flight) return 0;
  const planes = state.planes.filter(p => flight.planeIds.includes(p.id) && p.phase === 'ready');
  const order: AirOrder = requestedOrder ?? (planes[0].role === 'fighter' ? { kind: 'defend' } : { kind: 'attack', targetId: target?.motion.id ?? '' });
  // Ready aircraft have already refuelled, so their previous sortie age is irrelevant.
  if (order.kind === 'return' || !validAirOrder(actor, flight.id, planes.map(p => ({ ...p, flightTime: 0 })), order, actors)) return 0;
  flight.order = structuredClone(order); flight.notice = undefined;
  state.flights = state.flights.filter(f => f.id !== flight.id);
  state.flights.push(flight); state.flightSequence++;
  for (const plane of planes) {
    plane.sortie = (plane.sortie ?? 0) + 1;
    plane.phase = 'queued'; plane.flightId = flight.id; plane.targetId = order.kind === 'attack' ? order.targetId : undefined;
    plane.pilot = initialAirPilot(); plane.flightTime = 0; plane.timer = 0;
    plane.recoveryRequestedAt = undefined; plane.lossReason = undefined;
  }
  return planes.length;
}
/** One command path for a ready squadron's launch and an airborne squadron's retask. */
export function commandSquadron(actor: FleetActor, id: string, order: AirOrder, actors: FleetActor[]): boolean {
  const flight = squadronFlights(actor).find(f => f.id === id);
  if (!flight) return false;
  return activeFlight(flight, actor.airWing!.planes) ? orderFlight(actor, id, order, actors)
    : launchSquadron(actor, flight.squadronId, undefined, order, actors, id) > 0;
}
export function recallAircraft(actor: FleetActor, flightId?: string): void {
  for (const flight of actor.airWing?.flights ?? []) if (!flightId || flight.id === flightId) { flight.order = { kind: 'return' }; flight.notice = 'Recalled'; }
  for (const p of actor.airWing?.planes ?? []) {
    if (flightId && p.flightId !== flightId) continue;
    if (p.phase === 'queued') { p.phase = 'ready'; p.deckSlot = undefined; p.deckPosition = undefined; }
    else if (p.phase === 'taxi' || (p.phase === 'takeoff' && onFlightDeck(p))) p.phase = 'parking';
    else if (airborne(p) && p.phase !== 'landing') p.phase = 'returning';
  }
}
export function orderFlight(actor: FleetActor, flightId: string, order: AirOrder, actors: FleetActor[]): boolean {
  const flight = actor.airWing?.flights.find(f => f.id === flightId);
  if (!flight || !activeFlight(flight, actor.airWing!.planes) || physicalLoss(actor)) return false;
  if (order.kind === 'return') { recallAircraft(actor, flightId); return true; }
  const planes = actor.airWing!.planes.filter(p => p.flightId === flightId && ['queued', 'taxi', 'takeoff', 'outbound', 'attack', 'returning'].includes(p.phase));
  if (!planes.length || planes.some(p => p.flightTime > 470 || p.hp < 25)) return false;
  if (!validAirOrder(actor, flightId, planes, order, actors)) return false;
  flight.order = structuredClone(order); flight.notice = undefined;
  for (const p of planes) {
    p.targetId = order.kind === 'attack' ? order.targetId : undefined; p.pilot = initialAirPilot(); p.recoveryRequestedAt = undefined;
    if (['outbound', 'attack', 'returning'].includes(p.phase)) p.phase = 'outbound';
  }
  return true;
}
/** Oldest requests first, with emergency endurance ahead of ordinary traffic. */
export const AIR_TORPEDO = DEFAULT_AIR_TORPEDO;
export interface AirContext {
  seed?: number;
  actors: FleetActor[]; planes: Aircraft[]; shells: Shell[]; torpedoes: Torpedo[]; releases: AirRelease[];
  nextId: () => number; emit: (e: Omit<CombatEvent, 'sequence' | 'tick'>) => void;
}

File: crates/naval-sim/src/aviation/aircraft.rs
//! Carrier aircraft state and orders. Positions and mechanisms are CPU authoritative.
use crate::{definition::Vec3, rules::TeamId};
use serde::{Deserialize, Serialize};
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, ts_rs::TS)]
#[serde(tag = "kind", rename_all = "lowercase", deny_unknown_fields)]
pub enum AirOrder {
    #[serde(rename = "search-area")]
    SearchArea {
        center: [f64; 2],
        #[serde(rename = "radiusM")]
        radius_m: f64,
        altitude: SearchAltitude,
        policy: SearchPolicy,
    },
    Strike {
        #[serde(rename = "contactId")]
        contact_id: String,
    },
    #[serde(rename = "intercept-contact")]
    InterceptContact {
        #[serde(rename = "contactId")]
        contact_id: String,
    },
    Attack {
        #[serde(rename = "targetId")]
        target_id: String,
    },
    Patrol {
        point: Vec3,
    },
    Defend {
        #[serde(rename = "targetId")]
        #[ts(optional = nullable)]
        target_id: Option<String>,
    },
    Intercept {
        #[serde(rename = "flightId")]
        flight_id: String,
    },
    Escort {
        #[serde(rename = "flightId")]
        flight_id: String,
    },
    Return,
}
/// Search choices are operational tuning, not new aircraft capabilities.
#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "lowercase")]
pub enum SearchAltitude {
    Low,
    Medium,
    High,
}
impl SearchAltitude {
    pub(super) fn metres(self) -> f64 {
        match self {
            Self::Low => 200.0,
            Self::Medium => 850.0,
            Self::High => 1500.0,
        }
    }
}
#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "lowercase")]
pub enum SearchPolicy {
    Report,
    Shadow,
    Strike,
}
#[derive(Clone, Debug, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct SearchProgress {
    pub entry_position: Vec3,
    pub route: Vec<Vec3>,
    pub waypoint: usize,
    pub elapsed_seconds: f64,
    pub deadline_seconds: f64,
    pub shadow_seconds: f64,
    /// Actual flown samples, never a promise that nearby water is empty.
    pub trail: Vec<SearchSample>,
}
#[derive(Clone, Debug, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct SearchSample {
    pub position: Vec3,
    #[ts(type = "number")]
    pub tick: u64,
}
#[derive(Clone, Debug, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
#[ts(optional_fields)]
pub struct AirFlight {
    pub id: String,
    pub name: String,
    pub squadron_id: String,
    pub plane_ids: Vec<String>,
    pub order: AirOrder,
    pub notice: Option<String>,
    pub merged_into: Option<String>,
}
#[derive(Clone, Debug, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
#[ts(optional_fields)]
pub struct Aircraft {
    pub id: String,
    pub owner_id: String,
    pub team: TeamId,
    pub squadron_id: String,
    pub model_id: String,
    #[ts(type = "import('../../ships/blueprint').AircraftRole")]
    pub role: String,
    #[ts(as = "crate::frame_vocabulary::FlightPhase")]
    pub phase: String,
    pub position: Vec3,
    pub previous_position: Vec3,
    pub velocity: Vec3,
    pub heading: f64,
    pub pitch: f64,
    pub bank: f64,
    pub hp: f64,
    pub ammo: f64,
    pub payload: bool,
    pub wing_fold: f64,
    /// Managed routing datum, distinct from the fitted visual root.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub deck_datum: Option<Vec3>,
    #[serde(skip)]
    pub deck_local_attitude: Option<crate::aviation::aircraft_flight::FlightAttitude>,
    pub deck_position: Option<Vec3>,
    pub deck_heading: Option<f64>,
    pub timer: f64,
    pub flight_time: f64,
    pub cooldown: f64,
    pub target_id: Option<String>,
    /// Absent in a team frame: kills are scored from team reports there.
    #[ts(as = "Option<_>")]
    pub kills: u32,
    pub controls: crate::aviation::aircraft_flight::FlightControls,
    pub previous_controls: Option<crate::aviation::aircraft_flight::FlightControls>,
    pub previous_attitude: Option<crate::aviation::aircraft_flight::FlightAttitude>,
    /// Published as the pilot's visible activity alone (see
    /// [`AircraftBehavior`]); controller state stays private.
    #[ts(rename = "behavior", as = "Option<AircraftBehavior>")]
    pub pilot: AirPilot,
    pub deck_slot: Option<usize>,
    pub flight_id: Option<String>,
    pub recovery_requested_at: Option<f64>,
    pub loss_reason: Option<String>,
    pub navigation_target: Option<Vec3>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub search: Option<SearchProgress>,
    pub sortie: Option<u32>,
    pub wreck: Option<AirWreck>,
}
#[derive(Clone, Debug, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct AirWreck {
    pub age: f64,
    pub roll_rate: f64,
    pub impacted: bool,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AirPilot {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub maneuver: Option<crate::aviation::aircraft_tactics::FighterManeuver>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub formation: Option<crate::aviation::aircraft_formation::FormationState>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub defense: Option<crate::aviation::aircraft_defense::DefenseState>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recovery: Option<crate::aviation::aircraft_recovery::RecoveryProgress>,
    pub fire_discipline: Option<crate::aviation::air_gunnery::FireDiscipline>,
    pub think: f64,
    pub hostile_id: Option<String>,
    pub aim_time: f64,
    pub break_time: f64,
    pub break_cooldown: f64,
    pub break_point: Option<Vec3>,
    pub attack_heading: Option<f64>,
    pub attack_stage: Option<String>,
    pub attempts: u32,
    pub recovery_stage: Option<String>,
    pub recovery_side: Option<f64>,
    /// Seconds spent orbiting a lost strike report; bounds the search.
    #[serde(default, skip_serializing_if = "is_zero")]
    pub search_seconds: f64,
}
/// What the frame publishes of a pilot: the notices and manoeuvre a flight
/// line can show. The presentation filter writes it in the pilot's place
/// (`presentation::aircraft_behavior`); every other pilot field stays private.
#[derive(Clone, Debug, Default, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
#[ts(optional_fields)]
pub struct AircraftBehavior {
    pub recovery_notice: Option<String>,
    pub evasion_notice: Option<String>,
    pub maneuver: Option<String>,
}
fn is_zero(v: &f64) -> bool {
    *v == 0.0
}
#[derive(Clone, Debug, Default, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
#[ts(optional_fields)]
pub struct AirWingState {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recovery: Option<crate::aviation::air_recovery::CarrierRecovery>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub deck: Option<crate::aviation::deck_operations::DeckStatus>,
    pub planes: Vec<Aircraft>,
    pub launch_cooldown: f64,
    pub flights: Vec<AirFlight>,
    pub flight_sequence: u32,
    pub transfer_cooldown: f64,
}
#[derive(Clone, Debug, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
#[ts(optional_fields)]
pub struct AirRelease {
    #[ts(type = "number")]
    pub id: i64,
    pub owner_id: String,
    pub position: Vec3,
    pub velocity: Vec3,
    #[serde(skip_serializing_if = "Option::is_none")]
    /// Another team's weapon is cut to its visible dimensions and speed in a team
    /// frame (`team_view`); the owner's travels whole.
    #[ts(
        type = "Pick<import('../../ships/blueprint').TorpedoPart, 'diameterM' | 'lengthM' | 'speed'> & Partial<import('../../ships/blueprint').TorpedoPart>",
        optional
    )]
    pub weapon: Option<crate::definition::TorpedoPart>,
}
pub(super) const FIGHTER_AMMO_BURSTS: f64 = 16.0;
/// Assign a closed-set string in place. The stored bytes are identical to
/// `*slot = value.into()`; only the allocation is reused.
#[inline]
pub(super) fn set_str(slot: &mut String, value: &str) {
    if slot != value {
        slot.clear();
        slot.push_str(value);
    }
}
#[inline]
pub(super) fn set_opt_str(slot: &mut Option<String>, value: &str) {
    match slot {
        Some(existing) => set_str(existing, value),
        None => *slot = Some(value.to_owned()),
    }
}
pub(super) fn terminal_phase(phase: &str) -> bool {
    matches!(phase, "lost" | "withdrawn")
}
pub(super) fn airborne_phase(phase: &str) -> bool {
    matches!(
        phase,
        "takeoff" | "outbound" | "attack" | "returning" | "landing"
    )
}
pub fn terminal(p: &Aircraft) -> bool {
    terminal_phase(&p.phase)
}
pub fn airborne(p: &Aircraft) -> bool {
    airborne_phase(&p.phase)
}
pub(super) fn in_flight(p: &Aircraft) -> bool {
    airborne(p) && p.hp > 0.0
}
/// What a pilot is permitted to observe about another aircraft, borrowed
/// rather than cloned. Every field carries exactly the value the previous
/// `Aircraft` copy carried, so pilot decisions are unchanged; the fields the
/// pilot controllers never read are simply absent.
#[derive(Clone, Copy, Debug)]
pub(super) struct PlaneView<'a> {
    pub id: &'a str,
    pub flight_id: Option<&'a str>,
    pub hostile_id: Option<&'a str>,
    pub team: TeamId,
    pub role: &'a str,
    pub phase: &'a str,
    pub position: Vec3,
    pub velocity: Vec3,
    pub hp: f64,
    pub ammo: f64,
}
impl<'a> PlaneView<'a> {
    /// True airspeed; `velocity` is paced world motion, as on the aircraft itself.
    pub(super) fn airspeed(&self) -> f64 {
        crate::geometry::length(self.velocity) / crate::mobility::SHIP_PACE
    }
    pub(super) fn of(p: &'a Aircraft) -> Self {
        Self {
            id: &p.id,
            flight_id: p.flight_id.as_deref(),
            hostile_id: p.pilot.hostile_id.as_deref(),
            team: p.team,
            role: &p.role,
            phase: &p.phase,
            position: p.position,
            velocity: p.velocity,
            hp: p.hp,
            ammo: p.ammo,
        }
    }
    pub(super) fn in_flight(&self) -> bool {
        airborne_phase(self.phase) && self.hp > 0.0
    }
}
pub(crate) fn on_flight_deck(p: &Aircraft) -> bool {
    p.deck_slot.is_some()
        && matches!(
            p.phase.as_str(),
            "ready"
                | "queued"
                | "taxi"
                | "rollout"
                | "parking"
                | "rearming"
                | "raising"
                | "lowering"
                | "launch-ready"
        )
        || p.phase == "takeoff" && p.timer <= crate::aviation::aircraft_flight::TAKEOFF_ROLL_SECONDS
}
pub fn active_flight(f: &AirFlight, planes: &[Aircraft]) -> bool {
    planes.iter().any(|p| {
        p.flight_id.as_ref() == Some(&f.id)
            && !matches!(
                p.phase.as_str(),
                "ready"
                    | "rearming"
                    | "lost"
                    | "withdrawn"
                    | "hangar"
                    | "repairing"
                    | "raising"
                    | "lowering"
            )
    })
}
pub(super) fn aircraft_service_seconds(base: f64, hp: f64) -> f64 {
    base * (1.0 + (100.0 - hp.clamp(0.0, 100.0)) / 100.0)
}
impl Aircraft {
    /// True airspeed. `velocity` is world motion, which carries the world pace.
    pub(super) fn airspeed(&self) -> f64 {
        crate::geometry::length(self.velocity) / crate::mobility::SHIP_PACE
    }
    pub(super) fn forward(&self) -> Vec3 {
        [
            self.heading.sin() * self.pitch.cos(),
            self.pitch.sin(),
            -self.heading.cos() * self.pitch.cos(),
        ]
    }
}
pub fn create_air_wing(
    def: &crate::definition::ShipDefinition,
    owner_id: &str,
    team: TeamId,
    ground: &std::collections::BTreeMap<String, crate::aviation::aircraft_deck::GroundPose>,
) -> Option<AirWingState> {
    let wing = def.air_wing.as_ref()?;
    Some(AirWingState {
        recovery: None,
        planes: wing
            .squadrons
            .iter()
            .flat_map(|s| {
                (0..s.count as usize).map(move |i| Aircraft {
                    id: format!("{owner_id}/{}/{}", s.id, i + 1),
                    owner_id: owner_id.into(),
                    team,
                    squadron_id: s.id.clone(),
                    model_id: s.model_id.clone(),
                    role: s.role.clone(),
                    phase: "ready".into(),
                    position: [0.0; 3],
                    previous_position: [0.0; 3],
                    velocity: [0.0; 3],
                    heading: 0.0,
                    pitch: 0.0,
                    bank: 0.0,
                    hp: 100.0,
                    ammo: if s.role == "fighter" {
                        FIGHTER_AMMO_BURSTS
                    } else {
                        0.0
                    },
                    payload: s.role != "fighter",
                    wing_fold: if ground[&s.model_id].folding_wings {
                        1.0
                    } else {
                        0.0
                    },
                    deck_datum: None,
                    deck_local_attitude: None,
                    deck_position: None,
                    deck_heading: None,
                    timer: 0.0,
                    flight_time: 0.0,
                    cooldown: 0.0,
                    target_id: None,
                    kills: 0,
                    controls: Default::default(),
                    previous_controls: None,
                    previous_attitude: None,
                    pilot: Default::default(),
                    deck_slot: None,
                    flight_id: None,
                    recovery_requested_at: None,
                    loss_reason: None,
                    navigation_target: None,
                    search: None,
                    sortie: None,
                    wreck: None,
                })
            })
            .collect(),
        ..Default::default()
    })
}
